src/template.c

```diff
@@ -7335,6 +7292,83 @@ bool TemplateInstance::hasNestedArgs(Objects *args, bool isstatic)
     return nested != 0;
 }
 
+/*****************************************
+ * Append 'this' to the enclosing module members[] so the semantic routines
+ * will get called on the instance members.
+ */
+Dsymbols *TemplateInstance::appendToModuleMember()
+{
+    Module *md = tempdecl->getModule();
+
+    Module *mi = minst;
+    if (mi)
+    {
+        if (mi->isRoot())
+        {
+            /* If both modules mi and md where the template is declared are
+             * roots, instead use md to store the generated code in it.
+             */
+            if (md && md->isRoot())
+                mi = md;
```

Comment:
This is not necessary to fix the regression, but it will increase stability for the position where the instance will be inserted. For example:

a.d
```d
import ...;

template X() { ... }

X!()
```

b.d
```d
import a, ...;

X!()
```

With the command line `dmd -c a.d b.d`:

- If this adjustment does not exist, the instance X!() will be inserted to the member of *either* `a` or `b`. The actual position will depend on the semantic order.
- If this adjustment exists, the instance X!() will be *always* inserted to the member of `a`.

By that, the `X!()` code will be always stored into `a.obj`, in that case. I think the stable algorithm result is better than unstable, if there's not so much performance difference.